doozerlib/cli/get_nightlies.py

```diff
@@ -107,16 +219,277 @@ def rc_api_url(tag, arch):
+
+
+class EquivalenceSet:
+    """
+    Class to represent a set of nightlies from each arch that are equivalent
+    according to their EquivalenceBase. They can be sorted by creation date
+    (latest timestamp of nightlies in the set)
+    """
+
+    def __init__(self, nightly_for_arch: Dict[str, Dict]):
+        self.nightly_for_arch = nightly_for_arch
+        self.timestamp = max(nightly["releaseInfo"]["config"]["created"] for nightly in nightly_for_arch.values())
+
+    def generate_equivalents_with(self, arch: str, nightlies: List[Dict]) -> List['EquivalenceSet']:
+        """
+        Test each nightly for equivalence with all existing members, and return a list
+        of new sets extended with those that match.
+        """
+        if arch in self.nightly_for_arch:
+            raise EqSetDuplicateArchException(f"Cannot add arch {arch} when already present in {self}")
+
+        new_sets: List[EquivalenceSet] = []
+        for nightly in nightlies:
+            nightly_for_arch = {arch: nightly}  # initialize new set with candidate nightly
+            for exarch, existing in self.nightly_for_arch.items():
+                logger.debug(f"comparing {nightly['name']} and {existing['name']}")
+                if nightly["equivalence"] != existing["equivalence"]:
+                    break  # not equivalent
+                nightly_for_arch[exarch] = existing
+
+            if len(nightly_for_arch) > len(self.nightly_for_arch):  # all were added
+                new_sets.append(EquivalenceSet(nightly_for_arch))
+
+        return new_sets
+
+    def __repr__(self):
+        return " ".join(nightly['name'] for nightly in self.nightly_for_arch.values())
+
+    def details(self) -> str:
+        return f"{self}\n" + "".join(
+            f"{arch}: {nightly['name']} {nightly['phase']} rhcos: {nightly['releaseInfo']['displayVersions']['machine-os']['Version']}\n"
+            for arch, nightly in self.nightly_for_arch.items()
+        )
+
+    async def populate_deeper_equivalence(self, runtime):
+        """Prepare EquivalenceBases for deeper (more expensive) comparison"""
+        await asyncio.gather(*(
+            nightly["equivalence"].populate_deeper_equivalence(runtime, arch)
+            for arch, nightly in self.nightly_for_arch.items()
+        ))
+
+    def deeper_equivalence(self) -> bool:
+        """Check that all EquivalenceBases have deeper equivalency"""
+        nightlies = list(self.nightly_for_arch.values())
+        while len(nightlies) > 1:
+            this = nightlies.pop()
+            for other in nightlies:
+                logger.debug(f"comparing {this['name']} and {other['name']}")
+                if not this["equivalence"].deeper_equivalence(other["equivalence"]):
+                    logger.debug(f"deeper equivalence failed for {self}")
+                    return False
+
+        return True
+
+
+def generate_equivalence_sets(nightlies_for_arch: Dict[str, List[Dict]]) -> List[EquivalenceSet]:
```

Comment:
Think this should accept limit, we don't need to process all sets if we're asking for the first accepted one